include/cetl/span.h

```diff
@@ -1 +1,772 @@
+        : data_(arr.data())
+        , size_(ArrayLen)
+    {
+    }
+
+    /// Creates a span over an entire `const std::array` starting at the array's first element and with a size set to
+    /// `ArrayLen`.
+    /// @tparam ArrayElementType Deduced array element type to support SFNAE enablement.
+    /// @tparam ArrayLen         The length of the array. span::size() is set to this value.
+    /// @tparam type    Enables this override only if the array's element type is the same as the span's or if the
+    ///                 conversion is a simple [qualification
+    ///                 conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Qualification_conversions).
+    /// @param arr      The array to create the view into.
+    /// @see std::span::span()
+    template <
+        typename ArrayElementType,
+        std::size_t ArrayLen,
+        typename std::enable_if<std::is_convertible<const ArrayElementType (*)[], T (*)[]>::value, bool>::type = true>
+    constexpr span(const std::array<ArrayElementType, ArrayLen>& arr) noexcept
+        : data_(arr.data())
+        , size_(ArrayLen)
+    {
+    }
+
+    // TODO: C++20 range constructor.
+    // template< class R >
+    // explicit(extent != std::dynamic_extent)
+    // constexpr span( R&& range );
+
+    /// Copy constructor to create a span from another span. This overload allows conversion of any span to a dynamic
+    /// span.
+    /// @tparam DeducedElementType  The element type of the source span.
+    /// @tparam DeducedExtent       The extent of the source span.
+    /// @tparam type                Enables this override only if the element conversion is a simple [qualification
+    ///                 conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Qualification_conversions).
+    /// @param source   The span to copy from. The resulting span has `size() == source.size()` and
+    ///                 `data() == source.data()`
+    /// @see std::span()
+    template <class DeducedElementType,
+              size_type DeducedExtent,
+              typename std::enable_if<std::is_convertible<DeducedElementType (*)[], element_type (*)[]>::value,
+                                      bool>::type = true>
+    constexpr span(const span<DeducedElementType, DeducedExtent>& source) noexcept
+        : data_(source.data())
+        , size_(source.size())
+    {
+    }
+
+    /// Default copy constructor
+    /// @see std::span::span()
+    constexpr span(const span&) noexcept = default;
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Operators
+    // +----------------------------------------------------------------------+
+    /// @copydoc span::operator=()
+    ///
+    constexpr span& operator=(const span& rhs) noexcept = default;
+
+    /// @copydoc span::operator[]()
+    ///
+    constexpr reference operator[](size_type idx) const
```

Comment:
should be treat the data referenced in the span as const too?

Reply:
So, this is weird, but the standard defines c[begin/end] and cr[begin/end] but neither gcc nor clang actually implement this nor does cppreference.com list it. Otherwise, this operator is defined per the standard and I don't see any const reference version in it nor in cppreference.com.

Reply:
Resolving this as the implementation is per ISO/IEC specification
<img width="479" alt="image" src="https://user-images.githubusercontent.com/19180861/224227254-1abeec64-6a68-4587-a71f-b798b8fd6a5f.png">